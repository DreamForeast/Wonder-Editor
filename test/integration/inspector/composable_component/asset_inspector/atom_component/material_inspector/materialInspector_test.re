open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

open Js.Promise;

let _ =
  describe("material inspector", () => {
    let sandbox = getSandboxDefaultVal();

    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initState(~sandbox, ());
      InspectorCanvasTool.prepareInspectorEngineState(sandbox);

      CanvasTool.stubMainCanvasAndInspectorCanvasDom(~sandbox, ()) |> ignore;
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test rename material", () => {
      test(
        "if type is basicMaterial, rename to default basic material name should not work",
        () => {
          ConsoleTool.notShowMessage();

          let addedMaterialNodeId = MaterialAssetTool.addOneBasicMaterial();

          let newName = MainEditorMaterialTool.getDefaultBasicMaterialName();

          AssetInspectorTool.Rename.renameAssetMaterialNode(
            ~nodeId=addedMaterialNodeId,
            ~name=newName,
            (),
          );

          MainEditorAssetChildrenNodeTool.selectMaterialNode(
            ~nodeId=addedMaterialNodeId,
            (),
          );

          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          )
          |> ReactTestTool.createSnapshotAndMatch;
        },
      );
      test(
        "if type is lightMaterial, rename to default light material name should not work",
        () => {
          ConsoleTool.notShowMessage();

          let addedMaterialNodeId = MaterialAssetTool.addOneLightMaterial();

          let newName = MainEditorMaterialTool.getDefaultLightMaterialName();

          AssetInspectorTool.Rename.renameAssetMaterialNode(
            ~nodeId=addedMaterialNodeId,
            ~name=newName,
            (),
          );

          MainEditorAssetChildrenNodeTool.selectMaterialNode(
            ~nodeId=addedMaterialNodeId,
            (),
          );

          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          )
          |> ReactTestTool.createSnapshotAndMatch;
        },
      );

      describe("fix bug", () => {
        test("if rename to the same name, should warn", () => {
          ConsoleTool.notShowMessage();

          let warn =
            createMethodStubWithJsObjSandbox(
              sandbox,
              ConsoleTool.console,
              "warn",
            );

          let addedMaterialNodeId = MaterialAssetTool.addOneLightMaterial();

          let editorState = StateEditorService.getState();
          let engineState = StateEngineService.unsafeGetState();

          let {materialComponent, type_}: NodeAssetType.materialNodeData =
            OperateTreeAssetEditorService.unsafeFindNodeById(
              addedMaterialNodeId,
              editorState,
            )
            |> MaterialNodeAssetService.getNodeData;

          let newName =
            NodeNameAssetLogicService.getMaterialNodeName(
              ~material=materialComponent,
              ~type_,
              ~engineState,
            );

          AssetInspectorTool.Rename.renameAssetMaterialNode(
            ~nodeId=addedMaterialNodeId,
            ~name=newName,
            (),
          );

          warn |> expect |> toCalledOnce;
        });
        test("if rename to the existed name in the same dir, should fail", () => {
          let addedMaterialNodeId1 = MainEditorAssetIdTool.getNewAssetId();
          MainEditorAssetHeaderOperateNodeTool.addMaterial();
          let addedMaterialNodeId2 = MainEditorAssetIdTool.getNewAssetId();
          MainEditorAssetHeaderOperateNodeTool.addMaterial();

          let newName = "materialName";
          AssetInspectorTool.Rename.renameAssetMaterialNode(
            ~nodeId=addedMaterialNodeId1,
            ~name=newName,
            (),
          );
          let material2OldName =
            MainEditorAssetMaterialNodeTool.getMaterialName(
              ~nodeId=addedMaterialNodeId2,
              (),
            );
          AssetInspectorTool.Rename.renameAssetMaterialNode(
            ~nodeId=addedMaterialNodeId2,
            ~name=newName,
            (),
          );

          (
            MainEditorAssetMaterialNodeTool.getMaterialName(
              ~nodeId=addedMaterialNodeId1,
              (),
            ),
            MainEditorAssetMaterialNodeTool.getMaterialName(
              ~nodeId=addedMaterialNodeId2,
              (),
            ),
          )
          |> expect == (newName, material2OldName);
        });
      });
    });

    describe("change material type", () =>
      test(
        "if source material has no gameObject, still dispose source material",
        () => {
        let addedMaterialNodeId1 = MainEditorAssetIdTool.getNewAssetId();
        MainEditorAssetHeaderOperateNodeTool.addMaterial();
        let materialComponent1 =
          MainEditorAssetMaterialNodeTool.getMaterialComponent(
            ~nodeId=addedMaterialNodeId1,
            (),
          );

        MainEditorAssetChildrenNodeTool.selectMaterialNode(
          ~nodeId=addedMaterialNodeId1,
          (),
        );
        MaterialInspectorTool.changeMaterialType(
          ~material=materialComponent1,
          ~sourceMaterialType=MaterialDataAssetType.LightMaterial,
          ~targetMaterialType=MaterialDataAssetType.BasicMaterial,
          ~materialNodeId=addedMaterialNodeId1,
          (),
        );

        LightMaterialToolEngine.isAlive(materialComponent1)
        |> StateLogicService.getEngineStateToGetData
        |> expect == false;
      })
    );
  });