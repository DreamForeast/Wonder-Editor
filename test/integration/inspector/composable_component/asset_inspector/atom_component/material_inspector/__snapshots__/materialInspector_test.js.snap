// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`material inspector test rename material if type is basicMaterial, rename to default basic material name should not work 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-asset"
  >
    <article
      className="wonder-material-inspector"
    >
      <h1>
        Material
      </h1>
      <hr />
      <article
        className="inspector-item"
      >
        <div
          className="item-header"
          title="material's name"
        >
          Name
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="NoName Material"
          />
        </div>
      </article>
      <article
        className="inspector-item"
      >
        <div
          className="item-header"
          title="material's type"
        >
          Type
        </div>
        <div
          className="item-content"
        >
          <select
            onChange={[Function]}
            value="0"
          >
            <option
              value="0"
            >
              basic_material
            </option>
            <option
              value="1"
            >
              light_material
            </option>
          </select>
        </div>
      </article>
      <article
        className="wonder-basic-material"
      >
        <article
          className="color-pick"
        >
          <div
            className="item-header"
            title="color"
          >
            Color
          </div>
          <div
            className="item-content item-color"
          >
            <div
              className="color-hex"
              onClick={[Function]}
              style={
                Object {
                  "background": "#ffffff",
                }
              }
            />
            <div
              className="color-select"
              onClick={[Function]}
            >
              <img
                src="./public/img/color.png"
              />
            </div>
          </div>
        </article>
      </article>
    </article>
  </article>
</article>
`;

exports[`material inspector test rename material if type is lightMaterial, rename to default light material name should not work 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-asset"
  >
    <article
      className="wonder-material-inspector"
    >
      <h1>
        Material
      </h1>
      <hr />
      <article
        className="inspector-item"
      >
        <div
          className="item-header"
          title="material's name"
        >
          Name
        </div>
        <div
          className="item-content"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="New Material"
          />
        </div>
      </article>
      <article
        className="inspector-item"
      >
        <div
          className="item-header"
          title="material's type"
        >
          Type
        </div>
        <div
          className="item-content"
        >
          <select
            onChange={[Function]}
            value="1"
          >
            <option
              value="0"
            >
              basic_material
            </option>
            <option
              value="1"
            >
              light_material
            </option>
          </select>
        </div>
      </article>
      <article
        className="wonder-light-material"
      >
        <article
          className="color-pick"
        >
          <div
            className="item-header"
            title="material's diffuse color"
          >
            Diffuse Color
          </div>
          <div
            className="item-content item-color"
          >
            <div
              className="color-hex"
              onClick={[Function]}
              style={
                Object {
                  "background": "#ffffff",
                }
              }
            />
            <div
              className="color-select"
              onClick={[Function]}
            >
              <img
                src="./public/img/color.png"
              />
            </div>
          </div>
        </article>
        <article
          className="selectAssetNode-item"
        >
          <div
            className="item-header"
            title="material's diffuse map"
          >
            Diffuse Map
          </div>
          <div
            className="item-content"
          >
            <div
              className="img"
              onClick={[Function]}
              style={
                Object {
                  "opacity": "1",
                }
              }
            >
              <div
                className="img-dragBg"
                onDragEnter={[Function]}
                onDragLeave={[Function]}
                onDragOver={[Function]}
                onDrop={[Function]}
              />
            </div>
            <button
              className="remove"
              onClick={[Function]}
            >
              Remove
            </button>
          </div>
        </article>
        <article
          className="inspector-item wonder-float-input"
        >
          <div
            className="item-header component-label"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            title="material's shininess"
          >
            Shininess
          </div>
          <div
            className="item-content"
          >
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="32"
            />
          </div>
        </article>
      </article>
    </article>
  </article>
</article>
`;
